domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentCalculation.kt

```diff
@@ -0,0 +1,118 @@
+package org.oppia.android.domain.question
+
+import org.oppia.android.app.model.FractionGrade
+import org.oppia.android.app.model.UserAssessmentPerformance
+import javax.inject.Inject
+import kotlin.math.max
+
+/**
+ * Private class that computes the state of the user's performance at the end of a practice session.
+ * This class can exist across multiple training session instances, but calling code is responsible
+ * for ensuring it is properly reset. This class is not thread-safe, so it is the calling code's
+ * responsibility to synchronize access to this class.
+ */
+internal class QuestionAssessmentCalculation constructor(
+  private val viewHintPenalty: Int,
+  private val wrongAnswerPenalty: Int,
+  private val maxScorePerQuestion: Int,
+  private val questionSessionMetrics: List<QuestionSessionMetrics>,
+  private val totalScore: MutableFractionGrade,
+  private val scorePerSkillMapping: MutableMap<String, MutableFractionGrade>
+) {
+  /** Calculate the user's overall score and score per skill for this practice session. */
+  private fun calculateScores() {
```

Comment:
Thinking about this more, how important is it to save the side effects of this function? I'm actually wondering if we can just compute the latest state without using mutable state like MutableFractionGrade since we only appear to be computing things at the end of the session.

You may notice a theme with my feedback: any places we can reduce mutable state will improve the safety of the system, and often reduce the complexity of the code.

Reply:
I think avoiding mutable state is a really good point and something I'll definitely consider more! Thanks for all the feedback. However, I'm not sure there is a good way to do this without MutableFractionGrade. We're computing all the scores from the questionSessionMetrics list given to us - each item in the list holds the metrics for one question. Different questions can have overlapping tagged skills, so we need to keep track of the mutable scores for each skill as we process all of questionSessionMetrics. 

If we wanted to directly compute the latest state for each skill and store it in the proto object, I think we would have to filter the questions by their linked skills for every skill we're testing in the session and then do the calculations that are currently in calculateScores(). I think this would be less efficient and intuitive than just processing the metrics for each question once and copying the values into the final proto object, but maybe there is a better approach?

Let me know what you think!

Reply:
I think there are generally two expected ways to approach this problem (unless I'm missing some fundamental requirements):
1. Each time a question is answered, determine what needs to be tracked & track that. Results are a filtered aggregate/combination of this data.
2. Each time a question is answered, we record that it was answered. When results are needed, we calculate everything in that moment.

You're right that mutable state is needed now, but I think the choices allow flexibility in storing the mutable state either at the stats level (option 1) or at the event level (option 2). Option 1 is more performant when aggregating since we do some work ahead of time, and option 2 is more performant when recording. We should pick whichever option reflects typical access patterns, and which simplifies the implementation.

I think the current approach is more of option 2 in aggregation (everything is happening in compute) but using the storage of option 1. I don't quite see why we need to store MutableFractionGrade when it's only being changed/used in the compute function. It seems a bit like it's memoization, but unless compute is being called a lot that probably isn't needed.

Now, stepping back a moment. Given that the class no longer is storing state in Dagger (& thus has nothing to reset across sessions--we can just create a new one), and it's intended that computeAll only be computed at the end of the session, this additional caching seems a lot less necessary. Avoiding it actually results in a lot of simplifications, I think. One alternative that follows this pattern might be something like (which also conveniently avoids mutable state):

```kotlin
fun computePerformance(questionSessionMetrics: List<QuestionSessionMetrics>): UserAssessmentPerformance {
  val allQuestionScores: List<Int> = questionSessionMetrics.map { it.computeQuestionScore() }
  val totalPointsScored = allQuestionScores.sum()
  val totalAvailablePoints = allQuestionScores.size * maxScorePerQuestion

  // Note to Jessica: this also avoids needing to pass in the list to the class's constructor since we no longer need to initialize the list ahead of time. I think that results in a simpler API even though it comes at the expense of calling computeQuestionScore multiple times (thought that *could* be memoized if we found it to have performance issues).
  val scoresPerSkill: Map<String, List<Int>> = questionSessionMetrics.flatMap { questionMetric ->
    // Convert to List<Pair<String, QuestionSessionMetrics>>>
    questionMetric.linkedSkillIdsList.map { skillId -> skillId to questionMetric}
  }.groupBy(
    // Covert to Map<String, List<QuestionSessionMetrics>> for each linked skill ID.
    { (skillId, _) -> skillId },
    valueTransform = { (_, questionMetric) -> questionMetric }
  ).mapValues { (_, questionMetrics) ->
    // Compute the question score for each question, converting type to Map<String, List<Int>>
    questionMetrics.map { it.computeQuestionScore() }
  }

  // Assemble & return the proto, plus any other needed bookkeeping. Use the same sum()/size() * max to compute the total scores/available for each question metric.
}

private fun UserAssessmentPerformance.computeQuestionScore(): Int = if (!didViewSolution) {
  val hintsPenalty = numberOfHintsUsed * viewHintPenalty
  val wrongAnswerPenalty = getAdjustedWrongAnswerCount() * wrongAnswerPenalty
  (maxScorePerQuestion - hintsPenalty - wrongAnswerPenalty).coerceAtLeast(0)
} else 0

private fun UserAssessmentPerformance.getAdjustedWrongAnswerCount(): Int =
  (UserAssessmentPerformance - 1).coerceAtLeast(0)
```

(this is not complete & could be further split up if desired). https://stackoverflow.com/a/43614691 was a big help for the grouping part since I couldn't quite figure out how to map against the linked skill IDs.

WDYT?